Problem: Delayed Consul service failure notifications lead to Motr panic: !rep->hae_disconnected_previously

Situation, node-1 failed, failover in-progress to node-2

Hax services start..
```
Nov 05 22:29:42 sm28-r22.pun.seagate.com systemd[1]: Starting Cluster Controlled hare-hax-c2...
Nov 05 22:29:42 sm28-r22.pun.seagate.com systemd[1]: Starting Cluster Controlled hare-hax-c1...
Nov 05 22:29:42 sm28-r22.pun.seagate.com systemd[1]: Started Cluster Controlled hare-hax-c2.
Nov 05 22:29:42 sm28-r22.pun.seagate.com systemd[1]: Started Cluster Controlled hare-hax-c1.
Nov 05 22:29:42 sm28-r22.pun.seagate.com crmd[5501]:   notice: Transition aborted by status-2-hax-running doing modify hax-running=1: Transient attribute change
Nov 05 22:29:42 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:42,680 [DEBUG] {MainThread} Attempting to invoke the repeatable call: _get_motr_fids -> **hax-c2**
Nov 05 22:29:42 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:42,683 [DEBUG] {MainThread} Attempting to invoke the repeatable call: _get_motr_fids - **hax-c1**

Nov 05 22:29:42 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:42,722 [INFO] {MainThread} Welcome to HaX
Nov 05 22:29:43 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:43,440 [DEBUG] {qconsumer} Got EntrypointRequest(reply_context=139758571633632, req_id=0x0:0x1, remote_rpc_endpoint='192.168.0.3@o2ib:12345:1:1', process_fid=0x7200000000000001:0x6 -> **hax-c1 motr land entrypoint request**
Nov 05 22:29:43 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:43,450 [DEBUG] {qconsumer} Entrypoint request has been replied to
```
confd-c1 start initiated on node-2

```
Nov 05 22:29:47 sm28-r22.pun.seagate.com systemd[1]: Starting Cluster Controlled m0d@0x7200000000000001:0x9...
```
hax-c1 already says that confd-c1 and ios-c1 started

```
Nov 05 22:29:47 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:47,139 [DEBUG] {qconsumer} Got BroadcastHAStates(states=[HAState(fid=0x7200000000000001:0x9, status=OK), HAState(fid=0x7200000000000001:0x6, status=OK), HAState(fid=0x7200000000000001:0xc, status=OK), HAState(fid=0x7200000000000001:0x22, status=OK)], reply_to=None) message from queue
Nov 05 22:29:47 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:47,141 [INFO] {qconsumer} HA states: [HAState(fid=0x7200000000000001:0x9, status=OK), HAState(fid=0x7200000000000001:0x6, status=OK), HAState(fid=0x7200000000000001:0xc, status=OK), HAState(fid=0x7200000000000001:0x22, status=OK)]
Nov 05 22:29:47 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:47,142 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x9, status=OK), HAState(fid=0x7200000000000001:0x6, status=OK), HAState(fid=0x7200000000000001:0xc, status=OK), HAState(fid=0x7200000000000001:0x22, status=OK)] over ha_link

```
Stating confd-c2

```
Nov 05 22:29:48 sm28-r22.pun.seagate.com systemd[1]: Starting Cluster Controlled m0d@0x7200000000000001:0x52...
```

hax-c2 started,
```
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:48,512 [DEBUG] {MainThread} Attempting to invoke the repeatable call: get_leader_session
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:48,514 [DEBUG] {MainThread} The repeatable call succeeded: get_leader_session
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:48,519 [DEBUG] {MainThread} The repeatable call succeeded: _get_motr_fids
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:48,520 [INFO] {MainThread} Welcome to HaX
```

hax-c1 then sees self (hax-c1) and ios-c1 as failed for some reason from Consul catalog, probably due to a delay?

```
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,577 [DEBUG] {service-monitor} hax is now ServiceHealth.FAILED
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,579 [DEBUG] {service-monitor} ios is now ServiceHealth.FAILED
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,583 [DEBUG] {service-monitor} Changes in statuses: [HAState(fid=0x7200000000000001:0x6, status=FAILED), HAState(fid=0x7200000000000001:0xc, status=FAILED)]
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,593 [DEBUG] {qconsumer} Got BroadcastHAStates(states=[HAState(fid=0x7200000000000001:0x6, status=FAILED), HAState(fid=0x7200000000000001:0xc, status=FAILED)], reply_to=None) message from queue
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,593 [INFO] {qconsumer} HA states: [HAState(fid=0x7200000000000001:0x6, status=FAILED), HAState(fid=0x7200000000000001:0xc, status=FAILED)]
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,593 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x6, status=FAILED), HAState(fid=0x7200000000000001:0xc, status=FAILED)] over ha_link
```
and disconnects all the ha links,

```
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  67e0   WARN  [ha/halon/interface.c:336:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:6>
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  6700   WARN  [ha/ha.c:989:m0_ha_process_failed]  disconnecting local link

Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  67e0   WARN  [ha/halon/interface.c:336:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:c>
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  6690   WARN  [ha/halon/interface.c:523:halon_interface_link_is_disconnecting_cb]  hl=0x7f1a68002ac0 ep=192.168.0.3@o2ib:12345:1:1
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  65b0   WARN  [ha/halon/interface.c:1165:m0_halon_interface_disconnect]  hl=0x7f1a68002ac0 ep=192.168.0.3@o2ib:12345:1:1
```
Receives local entrypoint request from internal rconfc mostly, this leads to motr crash in hax-c1 as being not the first_request because process has not actually restarted but already running but halink is already disconnected and the entrypoint request is replied with `rep->hae_disconnected_previously = true` but the cookie is the same, see the following code,
```
static void ha_link_handle(struct m0_ha                       *ha,
                           const struct m0_uint128            *req_id,
                           const struct m0_ha_entrypoint_req  *req,
                           struct m0_ha_entrypoint_rep        *rep,
                           struct m0_ha_link                 **hl_ptr)
...
                } else if (m0_ha_cookie_is_eq(&ha->h_cookie_local,
                                             &req->heq_cookie_expected)) {
                        M0_LOG(M0_DEBUG, "link had been disconnected "
                               "previously, a new one shouldn't be created");
                        rep->hae_disconnected_previously = true;
                        add_new_link = false;
```

This leads to the below panic.

```
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,611 [DEBUG] {Dummy-1} Received entrypoint request from remote endpoint '192.168.0.3@o2ib:12345:1:1', process fid = 0x7200000000000001:0x6 The request will be processed in another thread.

Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,804 [DEBUG] {qconsumer} Got EntrypointRequest(reply_context=139758571639168, req_id=0x0:0x3, remote_rpc_endpoint='192.168.0.3@o2ib:12345:1:1', process_fid=0x7200000000000001:0x6, git_rev='02f6c4f2a513daae85730d7b50b83a7ca711afcd', pid=278443, **is_first_request=0**) message from queue
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,804 [DEBUG] {qconsumer} Processing entrypoint request from remote endpoint '192.168.0.3@o2ib:12345:1:1', process fid 0x7200000000000001:0x6
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,811 [DEBUG] {qconsumer} Services "confd" received: [{'ID': '9d08efab-2900-344c-9bd5-0b2858d73029', 'Node': 'srvnode-1', 'Address': '192.168.0.3', 'Datacenter': 'dc1', 'TaggedAddresses': {'lan': '192.168.0.3', 'lan_ipv4': '192.168.0.3', 'wan': '192.168.0.3', 'wan_ipv4': '192.168.0.3'}, 'NodeMeta': {'consul-network-segment': ''}, 'ServiceKind': '', 'ServiceID': '9', 'ServiceName': 'confd', 'ServiceTags': [], 'ServiceAddress': '192.168.0.3@o2ib:12345:2', 'ServiceWeights': {'Passing': 1, 'Warning': 1}, 'ServiceMeta': {}, 'ServicePort': 1, 'ServiceEnableTagOverride': False, 'ServiceProxy': {'MeshGateway': {}, 'Expose': {}}, 'ServiceConnect': {}, 'CreateIndex': 5631, 'ModifyIndex': 5631}, {'ID': 'ed3ec1db-f2f8-36a2-b3d4-a39bdea0e1b1', 'Node': 'srvnode-2', 'Address': '192.168.0.4', 'Datacenter': 'dc1', 'TaggedAddresses': {'lan': '192.168.0.4', 'lan_ipv4': '192.168.0.4', 'wan': '192.168.0.4', 'wan_ipv4': '192.168.0.4'}, 'NodeMeta': {'consul-network-segment': ''}, 'ServiceKind': '', 'ServiceID': '82', 'ServiceName': 'confd', 'ServiceTags': [], 'ServiceAddress': '192.168.0.4@o2ib:12345:2', 'ServiceWeights': {'Passing': 1, 'Warning': 1}, 'ServiceMeta': {}, 'ServicePort': 3, 'ServiceEnableTagOverride': False, 'ServiceProxy': {'MeshGateway': {}, 'Expose': {}}, 'ServiceConnect': {}, 'CreateIndex': 11496, 'ModifyIndex': 11496}]
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,812 [DEBUG] {qconsumer} Passing the entrypoint reply to hax.c layer
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  7820   WARN  [ha/halon/interface.c:1074:m0_halon_interface_entrypoint_reply]  req_id=0:3 rc=0 confd_nr=2 confd_quorum=2 rm_fid=<7300000000000001:a> rm_eps=192.168.0.3@o2ib:12345:2:1
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: [107B blob data]
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,812 [DEBUG] {qconsumer} Entrypoint request has been replied to
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: 2020-11-05 22:29:48,812 [DEBUG] {qconsumer} Waiting for the next message
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: motr[278443]:  6a10  FATAL  [lib/assert.c:50:m0_panic]  panic: !rep->hae_disconnected_previously at ha_entrypoint_state_cb() (ha/ha.c:310)  [git: sage-base-1.0-73-g02f6c4f] /var/motr1/hax/m0trace.278443
Nov 05 22:29:48 sm28-r22.pun.seagate.com hare-hax[278443]: Motr panic: !rep->hae_disconnected_previously at ha_entrypoint_state_cb() ha/ha.c:310
```
Similar thing happens with confd-c2
hax-c2 received entrypoint request from confd-c2 here with `is_first_request=true`, note that confd-c2 was already in starting process,

```
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,043 [DEBUG] {Dummy-1} Received entrypoint request from remote endpoint '192.168.0.4@o2ib:12345:2:3', process fid = 0x7200000000000001:0x52 The request will be processed in another thread.
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,099 [DEBUG] {qconsumer} Got EntrypointRequest(reply_context=139810044132576, req_id=0x0:0x2, remote_rpc_endpoint='192.168.0.4@o2ib:12345:2:3', process_fid=0x7200000000000001:0x52, git_rev='02f6c4f2a513daae85730d7b50b83a7ca711afcd', pid=1125, is_first_request=1) message from queue
```
hax-c2 also broadcasts that service started based on consul catalog (based on systemd,

```
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,142 [DEBUG] {service-monitor} confd is now ServiceHealth.OK
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,194 [DEBUG] {service-monitor} Changes in statuses: [HAState(fid=0x7200000000000001:0x52, status=OK), HAState(fid=0x7200000000000001:0x4f, status=OK), HAState(fid=0x7200000000000001:0x55, status=OK), HAState(fid=0x7200000000000001:0x6b, status=OK)]
...
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,322 [DEBUG] {qconsumer} Got BroadcastHAStates(states=[HAState(fid=0x7200000000000001:0x52, status=OK), HAState(fid=0x7200000000000001:0x4f, status=OK), HAState(fid=0x7200000000000001:0x55, status=OK), HAState(fid=0x7200000000000001:0x6b, status=OK)], reply_to=None) message from queue
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,324 [INFO] {qconsumer} HA states: [HAState(fid=0x7200000000000001:0x52, status=OK), HAState(fid=0x7200000000000001:0x4f, status=OK), HAState(fid=0x7200000000000001:0x55, status=OK), HAState(fid=0x7200000000000001:0x6b, status=OK)]
Nov 05 22:29:49 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:49,324 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x52, status=OK), HAState(fid=0x7200000000000001:0x4f, status=OK), HAState(fid=0x7200000000000001:0x55, status=OK), HAState(fid=0x7200000000000001:0x6b, status=OK)] over ha_link
```
In between we also have lnet errors,

```
Nov 05 22:29:49 sm28-r22.pun.seagate.com kernel: LNet: 252680:0:(o2iblnd_cb.c:2858:kiblnd_check_reconnect()) 192.168.0.4@o2ib: reconnect (conn race), 12, 12, msg_size: 4096, queue_depth: 8/8, max_frags: 256/256
Nov 05 22:29:49 sm28-r22.pun.seagate.com kernel: LNet: 252680:0:(o2iblnd_cb.c:2858:kiblnd_check_reconnect()) Skipped 10 previous similar messages
Nov 05 22:29:49 sm28-r22.pun.seagate.com kernel: LNet: 149654:0:(o2iblnd_cb.c:507:kiblnd_rx_complete()) Rx from 192.168.0.4@o2ib failed: 5
```
Here hax-c2 detected confd-c2 as FAILED and broadcasts the same,

```
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,259 [DEBUG] {service-monitor} ios is now ServiceHealth.FAILED
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,277 [DEBUG] {service-monitor} s3service is now ServiceHealth.FAILED
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,278 [DEBUG] {service-monitor} Changes in statuses: [HAState(fid=0x7200000000000001:0x52, status=FAILED), HAState(fid=0x7200000000000001:0x55, status=FAILED), HAState(fid=0x7200000000000001:0x6b, status=FAILED)]
```
```
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,866 [DEBUG] {qconsumer} The repeatable call succeeded: wrapper
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,866 [DEBUG] {qconsumer} Waiting for the next message
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,867 [DEBUG] {qconsumer} Got BroadcastHAStates(states=[HAState(fid=0x7200000000000001:0x52, status=FAILED), HAState(fid=0x7200000000000001:0x55, status=FAILED), HAState(fid=0x7200000000000001:0x6b, status=FAILED)], reply_to=None) message from queue
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,867 [INFO] {qconsumer} HA states: [HAState(fid=0x7200000000000001:0x52, status=FAILED), HAState(fid=0x7200000000000001:0x55, status=FAILED), HAState(fid=0x7200000000000001:0x6b, status=FAILED)]
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,867 [DEBUG] {qconsumer} Broadcasting HA states [HAState(fid=0x7200000000000001:0x52, status=FAILED), HAState(fid=0x7200000000000001:0x55, status=FAILED), HAState(fid=0x7200000000000001:0x6b, status=FAILED)] over ha_link
```
This disconnects the halinks wit confd-c2
Meanwhile confd-c2 starts and hax broadcasts go later, disconnecting the halinks with confd-c2
```
Nov 05 22:29:50 sm28-r22.pun.seagate.com motr-server[1125]: Started
Nov 05 22:29:50 sm28-r22.pun.seagate.com systemd[1]: Started Cluster Controlled m0d@0x7200000000000001:0x52.
Nov 05 22:29:50 sm28-r22.pun.seagate.com motr-server[1125]: m0d: systemd READY notification successfull
Nov 05 22:29:50 sm28-r22.pun.seagate.com motr-server[1125]: Press CTRL+C to quit.

Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,892 [DEBUG] {qconsumer} Broadcast HA state complete with the following message_ids = [MessageId(0x7f2664002ac0, 3), MessageId(0x7f266400dbd0, 11)]
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,892 [DEBUG] {qconsumer} Waiting for the next message
```
```
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,892 [DEBUG] {qconsumer} Setting process status in KV: processes/0x7200000000000001:0x52:{"state": "M0_CONF_HA_PROCESS_STARTED"}
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: motr[278441]:  87e0   WARN  [ha/halon/interface.c:336:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:52>
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: motr[278441]:  87e0   WARN  [ha/halon/interface.c:336:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:55>
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: motr[278441]:  87e0   WARN  [ha/halon/interface.c:336:halon_interface_process_failure_check]  disconnecting all links to no_id=<7200000000000001:6b>
```
Here confd-c2 requests entrypoint again, this time its not the first request but hax fails to fetch data from Consul and returns EAGAIN. But this does not help as the damage is done, again due to missing halink but cookie being the same (see above code snippet from motr halink) entrypoint request is replied with `rep->hae_disconnected_previously=true` leading to below crash in confd-c2.

```
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,996 [DEBUG] {qconsumer} Got EntrypointRequest(reply_context=139810044132288, req_id=0x0:0x4, remote_rpc_endpoint='192.168.0.4@o2ib:12345:2:3', process_fid=0x7200000000000001:0x52, git_rev='02f6c4f2a513daae85730d7b50b83a7ca711afcd', pid=1125, **is_first_request=0**) message from queue
Nov 05 22:29:50 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:50,996 [DEBUG] {qconsumer} Processing entrypoint request from remote endpoint '192.168.0.4@o2ib:12345:2:3', process fid 0x7200000000000001:0x52
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:51,001 [ERROR] {qconsumer} Failed to get the data from Consul. Replying with EAGAIN error code.

Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:51,001 [ERROR] {qconsumer} Failed to get the data from Consul. Replying with EAGAIN error code.
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: Traceback (most recent call last):
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 298, in get_leader_session_no_wait
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: return str(leader['Session'])
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: KeyError: 'Session'
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: During handling of the above exception, another exception occurred:
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: Traceback (most recent call last):
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/motr/__init__.py", line 134, in send_entrypoint_request_reply
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: sess = prov.get_leader_session_no_wait()
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: File "/opt/seagate/cortx/hare/lib64/python3.6/site-packages/hax/util.py", line 301, in get_leader_session_no_wait
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: 'Could not get the leader from Consul')
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: hax.exception.HAConsistencyException
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: motr[278441]:  9820   WARN  [ha/halon/interface.c:1074:m0_halon_interface_entrypoint_reply]  req_id=0:4 rc=11 confd_nr=0 confd_quorum=0 rm_fid=<7300000000000001:a> rm_eps=(null)
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: [107B blob data]
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:51,003 [DEBUG] {qconsumer} Reply sent
Nov 05 22:29:51 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:29:51,003 [DEBUG] {qconsumer} Waiting for the next message
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: motr[01125]:  3910  FATAL  [lib/assert.c:50:m0_panic]  panic: !rep->hae_disconnected_previously at ha_entrypoint_state_cb() (ha/ha.c:310)  [git: sage-base-1.0-73-g02f6c4f] /var/motr2/m0d-0x7200000000000001:0x52/m0trace.1125
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: Motr panic: !rep->hae_disconnected_previously at ha_entrypoint_state_cb() ha/ha.c:310 (errno: 0) (last failed: none) [git: sage-base-1.0-73-g02f6c4f] pid: 1125  /var/motr2/m0d-0x7200000000000001:0x52/m0trace.1125
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: Motr panic reason: HA has already decided that this process has failed. There is no good reason to continue doing something, and there is no code yet to handle graceful shutdown in this case. Let's just terminate the process and let HA do it's job.
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(m0_arch_backtrace+0x2f)[0x7f03f83eb44f]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(m0_arch_panic+0xf3)[0x7f03f83eb633]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x32dce4)[0x7f03f83dbce4]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x30acd5)[0x7f03f83b8cd5]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x32f147)[0x7f03f83dd147]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x32f19a)[0x7f03f83dd19a]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x3c9641)[0x7f03f8477641]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x3077f1)[0x7f03f83b57f1]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x301fa1)[0x7f03f83affa1]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(m0_thread_trampoline+0x5e)[0x7f03f83e213e]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libmotr.so.1(+0x33e38d)[0x7f03f83ec38d]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libpthread.so.0(+0x7ea5)[0x7f03f7b97ea5]
Nov 05 22:29:51 sm28-r22.pun.seagate.com motr-server[1125]: /lib64/libc.so.6(clone+0x6d)[0x7f03f6c888dd]
```
Later confd-c2 tried to start again but for some reason hax-c2 wasn't able to fetch data from consul as leader election failed because consul leader is elected based on the confd and no local confd was starting,
```
Nov 05 22:31:26 sm28-r22.pun.seagate.com systemd[1]: Starting Cluster Controlled m0d@0x7200000000000001:0x52...
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: /sys/kernel/mm/transparent_hugepage/defrag: always madvise [never]
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: MOTR_M0D_EP: 192.168.0.4@o2ib:12345:2:3
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: MOTR_PROCESS_FID: 0x7200000000000001:0x52
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: MOTR_HA_EP: 192.168.0.4@o2ib:12345:1:2
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: MOTR_CONF_XC: /etc/motr/confd.xc
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: vm.max_map_count = 30000000
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: Service FID: m0d-0x7200000000000001:0x52
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: BE log size is not configured
Nov 05 22:31:26 sm28-r22.pun.seagate.com motr-server[20096]: + exec /usr/bin/m0d -e lnet:192.168.0.4@o2ib:12345:2:3 -A linuxstob:addb-stobs -f '<0x7200000000000001:0x52>' -T linux -S stobs -D db -m 524288 -q 512 -n 82 -k 2097152 -K 536870912 -p 134217728 -C 524288 -N 83886080 -s 2097152 -y 300 -Y 300 -E 512 -c /etc/motr/confd.xc -H 192.168.0.4@o2ib:12345:1:2 -U -r 1073741824
Nov 05 22:31:26 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:31:26,861 [DEBUG] {qconsumer} Got EntrypointRequest(reply_context=139810044195312, req_id=0x0:0x16e, remote_rpc_endpoint='192.168.0.4@o2ib:12345:1:2', process_fid=0x7200000000000001:0x4f, git_rev='02f6c4f2a513daae85730d7b50b83a7ca711afcd', pid=278441, is_first_request=0) message from queue
Nov 05 22:31:26 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:31:26,861 [DEBUG] {qconsumer} Processing entrypoint request from remote endpoint '192.168.0.4@o2ib:12345:1:2', process fid 0x7200000000000001:0x4f
Nov 05 22:31:26 sm28-r22.pun.seagate.com hare-hax[278441]: 2020-11-05 22:31:26,866 [ERROR] {qconsumer} Failed to get the data from Consul. Replying with EAGAIN error code.
Nov 05 22:31:26 sm28-r22.pun.seagate.com hare-hax[278441]: Traceback (most recent call last):
```
confd-c1 also could not start and was finally banned as it was not able to complete the entry point request from hax-c1 as consul leader election kept failing,
```
Nov 05 22:31:21 sm28-r22.pun.seagate.com motr-server[942]: motr[00942]:  7c00   WARN  [ha/entrypoint.c:563:ha_entrypoint_client_fom_tick]  rlk_rc=-110
...
```
```
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:  warning: Processing failed monitor of hax-c1 on srvnode-2: not running
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:  warning: Processing failed start of motr-confd-c1 on srvnode-2: unknown error
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:  warning: Processing failed start of motr-confd-c1 on srvnode-2: unknown error
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:  warning: Processing failed monitor of motr-confd-c2 on srvnode-2: not running
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:  warning: Forcing motr-confd-c1 away from srvnode-2 after 1000000 failures (max=1000000)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Recover    hax-c1                (                   srvnode-2 )
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Stop       motr-confd-c1         (                   srvnode-2 )   due to node availability
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Recover    motr-confd-c2         (                   srvnode-2 )
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Start      motr-ios-c2           (                   srvnode-2 )   due to unrunnable motr-confd-c1 start (blocked)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Start      io_path_health-c2     (                   srvnode-2 )   due to unrunnable motr-ios-c2 start (blocked)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Start      motr-free-space-mon   (                   srvnode-2 )   due to unrunnable motr-ios-c1 start (blocked)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Start      haproxy-c2            (                   srvnode-2 )   due to unrunnable one-or-more:pcs_rsc_set_s3server-c2-1_s3server-c2-2_s3server-c2-3_s3server-c2-4_s3server-c2-5_s3server-c2-6_s3server-c2-7_s3server-c2-8_s3server-c2-9_s3server-c2-10_s3server-c2-11-1 (blocked)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice:  * Start      s3backcons-c2         (                   srvnode-2 )   due to unrunnable one-or-more:pcs_rsc_set_s3server-c2-1_s3server-c2-2_s3server-c2-3_s3server-c2-4_s3server-c2-5_s3server-c2-6_s3server-c2-7_s3server-c2-8_s3server-c2-9_s3server-c2-10_s3server-c2-11 (blocked)
Nov 05 22:31:24 sm28-r22.pun.seagate.com pengine[5500]:   notice: Calculated transition 54, saving inputs in /var/lib/pacemaker/pengine/pe-input-55.bz2
Nov 05 22:31:24 sm28-r22.pun.seagate.com crmd[5501]:   notice: Initiating stop operation motr-confd-c1_stop_0 locally on srvnode-2
```

All this lead to failover problems of the cluster.

--- a/hax/hax/motr/__init__.py
+++ b/hax/hax/motr/__init__.py
@@ -20,15 +20,17 @@ import ctypes as c
 import logging
 from errno import EAGAIN
 from typing import Any, List
+from queue import Queue
 
 from hax.exception import ConfdQuorumException, RepairRebalanceException
-from hax.message import (EntrypointRequest, HaNvecGetEvent,
+from hax.message import (BroadcastHAStates, EntrypointRequest, HaNvecGetEvent,
                          ProcessEvent)
 from hax.motr.delivery import DeliveryHerald
 from hax.motr.ffi import HaxFFI, make_array, make_c_str
 from hax.types import (ConfHaProcess, Fid, FidStruct, FsStats, HaNote,
                        HaNoteStruct, HAState, MessageId, ObjT, ReprebStatus,
-                       StobIoqError, ServiceHealth)
+                       StobIoqError, ServiceHealth, m0HaProcessEvent,
+                       m0HaProcessType, HaLinkMessagePromise)
 from hax.util import ConsulUtil
 
 LOG = logging.getLogger('hax')
@@ -108,6 +110,36 @@ class Motr:
                   " '{}', process fid = {}".format(remote_rpc_endpoint,
                                                    str(process_fid)) +
                   ' The request will be processed in another thread.')
+        try:
+            cns = ConsulUtil()
+            if is_first_request and (not cns.is_proc_client(process_fid)):
+                # This is the first start of this process or the process has
+                # restarted.
+                # Let everyone know that the process has restarted so that
+                # they can re-establish their connections with the process.
+                # Disconnect all the halinks this process is any.
+                # Motr clients are filtered out as they are the initiators of
+                # the rpc connections to motr ioservices. Presently there's no
+                # need identified to report motr client restarts, Consul will
+                # anyway detect the failure and report the same so we exclude
+                # reporting the same during their first entrypoint request.
+                # But we need to do it for motr server processes.
+                q: Queue = Queue(1)
+                LOG.debug('first entrypoint request, broadcasting FAILED')
+                self.queue.put(
+                    BroadcastHAStates(
+                        states=[HAState(fid=process_fid,
+                                        status=ServiceHealth.FAILED)],
+                        reply_to=q))
+                ids: List[MessageId] = q.get()
+                LOG.debug('waiting for broadcast of %s ep: %s',
+                          ids, remote_rpc_endpoint)
+                self.herald.wait_for_all(HaLinkMessagePromise(ids))
+        except Exception:
+            pass
+
+        LOG.debug('enqueue entrypoint request for %s',
+                  remote_rpc_endpoint)
         self.queue.put(
             EntrypointRequest(
                 reply_context=reply_context,
@@ -195,6 +227,13 @@ class Motr:
                               chp_type=chp_type,
                               chp_pid=chp_pid,
                               fid=fid)))
+
+        if chp_type == m0HaProcessType.M0_CONF_HA_PROCESS_M0D:
+            if chp_event == m0HaProcessEvent.M0_CONF_HA_PROCESS_STARTED:
+                self.queue.put(
+                    BroadcastHAStates(
+                        states=[HAState(fid=fid, status=ServiceHealth.OK)],
+                        reply_to=None))
 
     def _stob_ioq_event_cb(self, fid, sie_conf_sdev, sie_stob_id, sie_fd,
                            sie_opcode, sie_rc, sie_offset, sie_size,

--- a/hax/hax/motr/delivery.py
+++ b/hax/hax/motr/delivery.py
@@ -69,14 +69,45 @@ class DeliveryHerald:
                       self.recently_delivered[promise])
             del self.recently_delivered[promise]
 
+    def wait_for_all(self,
+                     promise: HaLinkMessagePromise,
+                     timeout_sec: float = 30.0):
+        """
+        Blocks the current thread until all of the messages in
+        promise._ids are reported by Motr as delivered.
+
+        Raises NotDelivered exception when timeout_sec exceeds.
+        """
+        for msg in promise._ids:
+            condition = Condition()
+            with self.lock:
+                self.waiting_clients[promise] = condition
+
+            with condition:
+                LOG.debug('Blocking until %s is confirmed', promise)
+                condition.wait(timeout=timeout_sec)
+            with self.lock:
+                if promise not in self.recently_delivered:
+                    raise NotDelivered('None of message tags =' +
+                                       str(promise) +
+                                       '  were delivered to Motr')
+                msgid = self.recently_delivered[promise]
+                if msgid in promise:
+                    promise._ids.remove(msgid)
+                LOG.debug('Thread unblocked - %s just received',
+                          self.recently_delivered[promise])
+                del self.recently_delivered[promise]
+
     def notify_delivered(self, message_id: MessageId):
         # [KN] This function is expected to be called from Motr.
         with self.lock:
+            LOG.debug('notify waiting clients %s',
+                      self.waiting_clients.items())
             for promise, client in self.waiting_clients.items():
                 if message_id in promise:
                     LOG.debug('Found a waiting client for %s: %s', message_id,
                               promise)
+                    self.recently_delivered[promise] = message_id
                     with client:
                         client.notify()
-                    self.recently_delivered[promise] = message_id
                     return

--- a/hax/hax/motr/hax.c
+++ b/hax/hax/motr/hax.c
@@ -541,7 +541,21 @@ static void msg_is_delivered_cb(struct m0_halon_interface *hi,
 static void msg_is_not_delivered_cb(struct m0_halon_interface *hi,
 				    struct m0_ha_link *hl, uint64_t tag)
 {
-	M0_LOG(M0_DEBUG, "noop");
+	struct m0_fid *proc_fid = &hl->hln_conn_cfg.hlcc_rpc_service_fid;
+
+	PyGILState_STATE gstate;
+	gstate = PyGILState_Ensure();
+
+	/* Notify hax about delivery failure. */
+	M0_LOG(M0_DEBUG, "msg not delivered, tag=%"PRIu64,
+		hl->hln_tag_broadcast_delivery);
+
+	PyObject *py_fid = toFid(proc_fid);
+	PyObject_CallMethod(hc0->hc_handler, "_msg_delivered_cb", "(OsKK)",
+			    py_fid, hl->hln_conn_cfg.hlcc_rpc_endpoint,
+			    hl->hln_tag_broadcast_delivery, hl);
+	Py_DECREF(py_fid);
+	PyGILState_Release(gstate);
 }
 
 static void link_connected_cb(struct m0_halon_interface *hi,

--- a/hax/hax/server.py
+++ b/hax/hax/server.py
@@ -54,13 +54,25 @@ def to_ha_states(data: Any) -> List[HAState]:
     if not data:
         return []
 
-    def get_status(checks: List[Dict[str, Any]]) -> ServiceHealth:
-        ok = all(x.get('Status') == 'passing' for x in checks)
-        return ServiceHealth.OK if ok else ServiceHealth.FAILED
+    def get_svc_node(checks: List[Dict[str, Any]], svc_id: str) -> str:
+        for x in checks:
+            if x.get('ServiceID') == svc_id:
+                return str(x.get('Node'))
+        return ""
+
+    def get_svc_health(item: Any) -> ServiceHealth:
+        node = get_svc_node(item['Checks'], item['Service']['ID'])
+        LOG.debug('Checking current state of the process %s',
+                  item['Service']['ID'])
+        status: ServiceHealth = ConsulUtil().get_service_health(
+                                               item['Service']['Service'],
+                                               node,
+                                               int(item['Service']['ID']))
+        return status
 
     return [
         HAState(fid=create_process_fid(int(t['Service']['ID'])),
-                status=get_status(t['Checks'])) for t in data
+                status=get_svc_health(t)) for t in data
     ]
 
 

--- a/hax/hax/types.py
+++ b/hax/hax/types.py
@@ -216,3 +216,41 @@ class ServiceHealth(Enum):
 
 
 HAState = NamedTuple('HAState', [('fid', Fid), ('status', ServiceHealth)])
+
+
+class m0HaProcessEvent(Enum):
+    M0_CONF_HA_PROCESS_STARTING = 0
+    M0_CONF_HA_PROCESS_STARTED = 1
+    M0_CONF_HA_PROCESS_STOPPING = 2
+    M0_CONF_HA_PROCESS_STOPPED = 3
+
+    def str_to_Enum(self, evt):
+        events = {'M0_CONF_HA_PROCESS_STARTING':
+                  self.M0_CONF_HA_PROCESS_STARTING,
+                  'M0_CONF_HA_PROCESS_STARTED':
+                  self.M0_CONF_HA_PROCESS_STARTED,
+                  'M0_CONF_HA_PROCESS_STOPPING':
+                  self.M0_CONF_HA_PROCESS_STOPPING,
+                  'M0_CONF_HA_PROCESS_STOPPED':
+                  self.M0_CONF_HA_PROCESS_STOPPED}
+        return events[evt]
+
+    def __repr__(self):
+        return self.name
+
+
+class m0HaProcessType(Enum):
+    M0_CONF_HA_PROCESS_OTHER = 0
+    M0_CONF_HA_PROCESS_KERNEL = 1
+    M0_CONF_HA_PROCESS_M0MKFS = 2
+    M0_CONF_HA_PROCESS_M0D = 3
+
+    def str_to_Enum(self):
+        types = {'M0_CONF_HA_PROCESS_OTHER': self.M0_CONF_HA_PROCESS_OTHER,
+                 'M0_CONF_HA_PROCESS_KERNEL': self.M0_CONF_HA_PROCESS_KERNEL,
+                 'M0_CONF_HA_PROCESS_M0MKFS': self.M0_CONF_HA_PROCESS_M0MKFS,
+                 'M0_CONF_HA_PROCESS_M0D': self.M0_CONF_HA_PROCESS_M0D}
+        return types[self.name]
+
+    def __repr__(self):
+        return self.name

--- a/hax/hax/util.py
+++ b/hax/hax/util.py
@@ -32,7 +32,8 @@ from requests.exceptions import RequestException
 from urllib3.exceptions import HTTPError
 
 from hax.exception import HAConsistencyException
-from hax.types import ConfHaProcess, Fid, FsStatsWithTime, ObjT
+from hax.types import (ConfHaProcess, Fid, FsStatsWithTime,
+                       ObjT, ServiceHealth)
 
 __all__ = ['ConsulUtil', 'create_process_fid', 'create_drive_fid']
 
@@ -300,29 +301,29 @@ class ConsulUtil:
             raise HAConsistencyException('Failed to communicate to'
                                          ' Consul Agent') from e
 
+    def get_svc_status(self, srv_fid: Fid) -> str:
+        try:
+            key = f'processes/{srv_fid}'
+            raw_data = self.kv.kv_get(key)
+            LOG.debug('Raw value from KV: %s', raw_data)
+            data = raw_data['Value']
+            value: str = json.loads(data)['state']
+            return value
+        except Exception:
+            return 'Unknown'
+
     def get_m0d_statuses(self) -> List[Tuple[ServiceData, str]]:
         """
         Return the list of all Motr service statuses according to Consul
         watchers. The following services are considered: ios, confd.
         """
-        def get_status(srv: ServiceData) -> str:
-            try:
-                key = f'processes/{srv.fid}'
-                raw_data = self.kv.kv_get(key)
-                LOG.debug('Raw value from KV: %s', raw_data)
-                data = raw_data['Value']
-                value: str = json.loads(data)['state']
-                return value
-            except Exception:
-                return 'Unknown'
-
         m0d_services = set(['ios', 'confd'])
         result = []
         for service_name in self._catalog_service_names():
             if service_name not in m0d_services:
                 continue
             data = self.get_service_data_by_name(service_name)
-            result += [(item, get_status(item)) for item in data]
+            result += [(item, self.get_svc_status(item.fid)) for item in data]
         return result
 
     def get_service_data_by_name(self, name: str) -> List[ServiceData]:
@@ -362,6 +363,33 @@ class ConsulUtil:
                 FidWithType(fid=mk_fid(ObjT.SERVICE, srv_fidk),
                             service_type=srv_type))
         return services
+
+    def is_proc_client(self, process_fid: Fid) -> bool:
+        node_items = self.kv.kv_get('m0conf/nodes', recurse=True)
+        fidk = str(process_fid.key)
+
+        # This is the RegExp to match the keys in Consul KV that describe
+        # the Motr services that are enclosed into the Motr process that has
+        # the given fidk.
+        # We filter out motr client entries to check if the given process fid
+        # corresponds to a motr client or server process.
+        #
+        # Note: we assume that fidk uniquely identifies the given process
+        # within the whole cluster (that's why we are not interested in the
+        # hostnames here).
+        #
+        # Examples of the key that will match:
+        #   m0conf/nodes/srvnode-1/processes/39/services/m0_client_s3
+        regex = re.compile(
+            f'^m0conf\\/.*\\/processes\\/{fidk}\\/services\\/(.+)$')
+        for node in node_items:
+            match_result = re.match(regex, node['Key'])
+            if not match_result:
+                continue
+            srv_type = match_result.group(1)
+            if 'm0_client' in srv_type:
+                return True
+        return False
 
     def get_conf_obj_status(self, obj_t: ObjT, fidk: int) -> str:
         # 'node/<node_name>/process/<process_fidk>/service/type'
@@ -429,6 +457,15 @@ class ConsulUtil:
         LOG.debug('Setting process status in KV: %s:%s', key, data)
         self.kv.kv_put(key, data)
 
+    def get_process_status(self, event: ConfHaProcess) -> None:
+        assert 0 <= event.chp_event < len(ha_process_events), \
+            f'Invalid event type: {event.chp_event}'
+
+        data = json.dumps({'state': ha_process_events[event.chp_event]})
+        key = f'processes/{event.fid}'
+        LOG.debug('Setting process status in KV: %s:%s', key, data)
+        self.kv.kv_put(key, data)
+
     def drive_name_to_id(self, uid: str) -> str:
         drive_id = ''
         # 'm0conf/nodes/<node_name>/processes/<process_fidk>/disks/<disk_uuid>'
@@ -446,6 +483,32 @@ class ConsulUtil:
         key = f'process/{fid}'
         LOG.debug('Setting disk state in KV: %s:%s', key, data)
         self.kv.kv_put(key, data)
+
+    @repeat_if_fails()
+    def get_service_health(self, service_name: str,
+                           node: str, svc_id: int) -> ServiceHealth:
+
+        try:
+            node_data: List[Dict[str, Any]] = self.cns.health.node(node)[1]
+            LOG.debug('Node Data: %s', node_data)
+            status = ServiceHealth.UNKNOWN
+            for item in node_data:
+                if (item['ServiceName'] == service_name and
+                        item['ServiceID'] == str(svc_id)):
+                    if item['Status'] == 'passing':
+                        status = ServiceHealth.OK
+                    elif item['Status'] == 'warning':
+                        fid = create_process_fid(svc_id)
+                        svc_consul_status = self.get_svc_status(fid)
+                        if svc_consul_status in ('M0_CONF_HA_PROCESS_STARTING',
+                                                 'M0_CONF_HA_PROCESS_STARTED'):
+                            status = ServiceHealth.OK
+                    else:
+                        status = ServiceHealth.FAILED
+        except (ConsulException, HTTPError, RequestException) as e:
+            raise HAConsistencyException('Failed to communicate '
+                                         'to Consul Agent') from e
+        return status
 
 
 def dump_json(obj) -> str:

